### PR TITLE
Allow filesystem commands to access files with glob metachars in name

### LINKED ISSUE #10498: `glob` command supports more metacharacters than filesystem commands like `cp` and `mv`
### Describe the bug

It looks like only the `glob` command supports the the csh-style `{}' metacharacters, the file utilities do not.

You could argue this is more of an enhancement request, but I think globbing is so fundamental a shell experience that it must be consistent across use cases. 

### How to reproduce

```
> touch a.foo a.bar
> ls
╭───┬───────┬──────┬──────┬──────────╮
│ # │ name  │ type │ size │ modified │
├───┼───────┼──────┼──────┼──────────┤
│ 0 │ a.bar │ file │  0 B │ now      │
│ 1 │ a.foo │ file │  0 B │ now      │
╰───┴───────┴──────┴──────┴──────────╯
> glob a.{foo,bar}
╭───┬────────────────────────────────────╮
│ 0 │ /home/bobhy/src/rust/work/r3/a.bar │
│ 1 │ /home/bobhy/src/rust/work/r3/a.foo │
╰───┴────────────────────────────────────╯
```
So `glob` supports `{}`.  Great!
```
> mkdir dir_a
> cp a.{foo,bar} dir_a
Error:   × No matches found
   ╭─[entry #187:1:1]
 1 │ cp a.{foo,bar} dir_a
   ·    ─────┬─────
   ·         ╰── no matches found
   ╰────

> ucp a.{foo,bar} dir_a
Error: nu::shell::file_not_found

  × File not found
   ╭─[entry #188:1:1]
 1 │ ucp a.{foo,bar} dir_a
   ·     ─────┬─────
   ·          ╰── file not found
   ╰────
```
`cp` and `ucp` do not.  Not shown here, but neither do `rm`, `mv`, `touch` .

### Expected behavior

I'd really like all Nu globbing to support '{}' metacharacters.

### Screenshots

_No response_

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.85.1                                   |
| branch             | main                                     |
| commit_hash        | 95a4649cc9933045d275d508f33676da16159d18 |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.70.0 (90c541806 2023-05-31)      |
| rust_channel       | 1.70.0-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.70.0 (ec8a8a0ca 2023-04-25)      |
| build_time         | 2023-09-24 10:42:53 -07:00               |
| build_rust_channel | release                                  |
| allocator          | mimalloc                                 |
| features           | default, sqlite, trash, which, zip       |
| installed_plugins  |                                          |

### Additional context

Looking at the code, the filesystem utilities use crates/nu-glob, but `glob` uses `wax::Glob`.

Understood, I could do:
```
glob a.{foo,bar} | each {|p| cp $p dir_a}
```
with a minor loss of efficiency.  But that's just papering over the crack...

### PR DESCRIPTION
**Pull request by bobhy**
MEGA-EDIT:
Completely changing goals of this PR! 
Work to date has found a fix for a good bug, but the original idea of supporting `{}` brace expansion like `bash` no longer seems achievable as a globbing enhancement. So this PR is now all about:

Fix problem #10571.
# Description

before:  Args to `cp`, `mv`, `rm` are assumed to be globs within each command.  That means a file that looks like a glob is not matched by the glob  and takes some thinking about to access.  That is, the glob pattern shadows a file with the same spelling.

```
> ls a*
╭───┬────────┬──────┬──────┬────────────────╮
│ # │  name  │ type │ size │    modified    │
├───┼────────┼──────┼──────┼────────────────┤
│ 0 │ a[bc]d │ file │  0 B │ 34 seconds ago │
│ 1 │ abd    │ file │  0 B │ now            │
│ 2 │ acd    │ file │  0 B │ now            │
╰───┴────────┴──────┴──────┴────────────────╯

> cp --verbose 'a[bc]d' dest
copied /home/bobhy/src/rust/work/r4/abd to /home/bobhy/src/rust/work/r4/dest/abd
copied /home/bobhy/src/rust/work/r4/acd to /home/bobhy/src/rust/work/r4/dest/acd

> ## Note -- a[bc]d *not* copied, and seemingly hard to access.
> cp --verbose 'a\[bc\]d' dest
Error:   × No matches found
   ╭─[entry #33:1:1]
 1 │ cp --verbose 'a\[bc\]d' dest
   ·              ─────┬────
   ·                   ╰── no matches found
   ╰────

> #.. but not completely inaccessible.
> cp --verbose 'a[[]bc[]]d' dest
copied /home/bobhy/src/rust/work/r4/a[bc]d to /home/bobhy/src/rust/work/r4/dest/a[bc]d

```
Before_2: if file has glob metachars but isn't a valid pattern, user gets a confusing error:
```
> touch 'a[b'
> cp 'a[b' dest
Error:   × Pattern syntax error near position 30: invalid range pattern
   ╭─[entry #13:1:1]
 1 │ cp 'a[b' dest
   ·    ──┬──
   ·      ╰── invalid pattern
   ╰────
```

After: Args to `cp`, `mv`, etc. are tried first as literal files, and only as globs if not found to be files.
```
> cp --verbose 'a[bc]d' dest
copied /home/bobhy/src/rust/work/r4/a[bc]d to /home/bobhy/src/rust/work/r4/dest/a[bc]d
> cp --verbose '[a][bc]d' dest
copied /home/bobhy/src/rust/work/r4/abd to /home/bobhy/src/rust/work/r4/dest/abd
copied /home/bobhy/src/rust/work/r4/acd to /home/bobhy/src/rust/work/r4/dest/acd
```

After_2: file with glob metachars but invalid pattern just works (though the name probably requires string quotes)
```
> cp --verbose 'a[b' dest
copied /home/bobhy/src/rust/work/r4/a[b to /home/bobhy/src/rust/work/r4/dest/a[b
```

So, with this fix,  a file shadows a glob pattern with the same spelling.  You must invent some other spelling for the glob pattern to match the desired files.  I think that's less confusing to the user: if `ls` shows a file with a weird name, s/he'll still be able to copy, rename or delete it.  

;tl;dr
### Original title: Enable {} glob metachars in command arguments 

... To make `nu` feel more like `csh`.
Fixes #10498.

So far, only `ucp` command has been touched, as a proof of concept.  But we should proceed through the rest of the filesystem commands.

This PR uses crate `wax` for globbing instead of fixing the internal crate `nu_glob`.  Once all the clients are upgraded, `nu_glob` could be deprecated.

# Description
Enables glob alternation to be used in command line arguments (where arguments are supposed to refer to paths in filesystem).
e.g:
```
> ucp cr{awl,eep}.txt dest
> ls dest
╭───┬────────────────┬──────┬──────┬───────────────╮
│ # │      name      │ type │ size │   modified    │
├───┼────────────────┼──────┼──────┼───────────────┤
│ 0 │ dest/crawl.txt │ file │  0 B │ 2 minutes ago │
│ 1 │ dest/creep.txt │ file │  0 B │ 2 minutes ago │
╰───┴────────────────┴──────┴──────┴───────────────╯
```
`{` and `}` are nu parser tokens, so must be quoted if they are not embedded within the glob.
```
> ucp '{crawl,hobble,sprint}.txt' dest
> ls dest
╭───┬─────────────────┬──────┬──────┬────────────────╮
│ # │      name       │ type │ size │    modified    │
├───┼─────────────────┼──────┼──────┼────────────────┤
│ 0 │ dest/crawl.txt  │ file │  0 B │ 11 seconds ago │
│ 1 │ dest/hobble.txt │ file │  0 B │ 11 seconds ago │
│ 2 │ dest/sprint.txt │ file │  0 B │ 11 seconds ago │
╰───┴─────────────────┴──────┴──────┴────────────────╯
```
Note that alternation works for more than 2 alternatives, as shown above.

# User-Facing Changes

Currently, the new metachars can only be used in `ucp` command.

All uses of the old metachars should give same results as they used to, so no breaking changes.
Small changes to error messages when a glob fails to match anything.  

# Tests + Formatting
- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting

### CODE REVIEW

**Comment by fdncred** [comment]
Our friendly neighborhood windows CI strikes again. Without looking further, I have 2 guesses why it's failing.
1. Paths are in double quotes so with "C:\path\to\file" all "\" become escapes.
2. There's a mixture of slashes with 'C:\path/to/file.txt'. 
I've seen both of these scenarios fail sometimes.

Generally, I think this PR is cool.

**Comment by bobhy** [comment]
I see a couple of problems.  Firing up my Windows dev environment, which is a performance hit all of its own...

File: crates/nu-cmd-base/src/arg_glob.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+// expand a glob into matching paths, using FS_CMD_WALK_BEHAVIOR
+//todo: see about returning WalkError or BuildError, or enabling miette and dealing with diagnostic
+pub fn arg_glob(
+    pattern: &Spanned<String>, // the glob
+    include_dirs: bool,        // include dirs in results.  Default (f) only include files
+    cwd: &PathBuf,             // current working directory
+) -> Result<Vec<PathBuf>, ShellError> {
+    // eventually, windows only, and with better performance
+    // transform something like: C:\Users\me\.../{abc,def}*.txt to C\:/Users/me/.../{abc,def}*.txt
```
Comment (fdncred) [inline crates/nu-cmd-base/src/arg_glob.rs:20]:
```suggestion
    // transform something like: C:\Users\me\.../{abc,def}*.txt to C:/Users/me/.../{abc,def}*.txt
```

File: crates/nu-cmd-base/src/arg_glob.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    // such a string has good chance of working as a rooted glob pattern on Windows
+    // User will have to quote metachars with [c] rather than \c.
+    let processed_pattern = pattern
+        .item
+        .replace('\\', "/")
+        .replace(r#":/"#, r#"\:/"#) // must come after above
+        .replace('(', r#"\("#) // when path includes Program Files (x86)
+        .replace(')', r#"\)"#);
```
Comment (fdncred) [inline crates/nu-cmd-base/src/arg_glob.rs:28]: I'm wondering what effect this will have on paths with spaces like `Program Files` and `Program Files (x86)`. I see that lots of things are being replaced here but I'm not sure how that affects things here with these particular paths.

File: crates/nu-cmd-base/src/arg_glob.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        // path we're trying to glob doesn't exist,
+        {
+            PathBuf::new() // ensure glob doesn't match anything
+        }
+        Err(e) => return Err(ShellError::ErrorExpandingGlob(format!("{e}"), pattern.span)),
+    };
+
+    let mut rv: Vec<PathBuf> = Vec::new();
+    for w in glob
+        .walk_with_behavior(path, FS_CMD_WALK_BEHAVIOR)
```
Comment (fdncred) [inline crates/nu-cmd-base/src/arg_glob.rs:49]: Since depth is usize::MAX, is there exposure here that we could have performance problems because it's searching for deeply nested matches?

File: crates/nu-cmd-base/src/arg_glob.rs (hunk 2 above)
Comment (bobhy) [reply crates/nu-cmd-base/src/arg_glob.rs]: Embedded spaces won't be affected by these changes.  A pattern or simple path with an embedded space would have to be enclosed in quotes anyway.  It has been tested and works with `C:\Program Files (x86)\*`!

File: crates/nu-cmd-base/src/arg_glob.rs (hunk 3 above)
Comment (bobhy) [reply crates/nu-cmd-base/src/arg_glob.rs:43]: Yes.  But it's the same depth currently used by `glob` command.  Does linux `cp` have any limit?

**Comment by jntrnr** [comment]
We have to be careful with adding glob syntax. Anything that's part of a valid filename on any of our platforms can't be used.

Can you confirm that `{` isn't part of a valid filename on Linux, Windows, and macOS?

**Comment by bobhy** [comment]
A key point of this change is to be able to use a simple file path or a glob pattern as argument to a filesystem command like `ucp`. That means that normal looking file paths have to be parsable as glob patterns (so they will match just one file).  That's a problem on Windows, were backslashes and colons in absolute (rooted) patterns will cause glob parse errors.  
This is a well-known issue with globs on Windows.  `nu_glob` has some fixes, but `wax` does not.  The .NET glob (written by Microsoft engineers) doesn't even try to support a rooted pattern.

This checkin includes a hack for path/pattern arguments on windows only that comvers many cases.
```
// Sanitize a glob pattern for use on windows.
// this allows filesystem commands like `cp` to accept an arg that can be a path or a glob.
// Deal with 3 aspects of windows paths that can confound glob patterns:
// * path separator: change `\` to `/`. Rust file API on windows supports this in folder separators and in UNC paths.
// * drive letter colon: change `C:` to `C\:`
// * quote parens (such as the infamous `Program Files (x86)`).
// But this means user cannot use `\` to quote other metachars.
// (Windows) users of glob should adopt the habit of quoting with single char classes, e.g instead of `\*`, use `[*]`.
// e.g transform something like: `C:\Users\me\.../{abc,def}*.txt` to `C\:/Users/me/.../{abc,def}*.txt`
//todo: investigate enhancing [wax] to support this more elegantly.
pub fn windows_pattern_hack(pat: &str) -> String {
    pat.replace('\\', "/")
        .replace(r#":/"#, r#"\:/"#) // must come after above
        .replace('(', r#"\("#) // when path includes Program Files (x86)
        .replace(')', r#"\)"#)
}
```
Windows users, let me know how this works for you!

File: crates/nu-cmd-base/src/arg_glob.rs (hunk 3 above)
Comment (fdncred) [reply crates/nu-cmd-base/src/arg_glob.rs:43]: I have no idea what the limitations of `cp` may be.

Another thing I was thinking was, since the depth is set to max, is there the possibility of false positive completions here? like you do something with the {} syntax, intending to change files in the current folder, but some 10 levels deeper there are things that match.

I'm not suggesting that you change the depth, I'm just trying to think if there are ways it can fail. I think I originally wrote glob and made it take the default max depth to begin with.

Comment (bobhy) [reply crates/nu-cmd-base/src/arg_glob.rs:43]: Seems the various OS filesystems have depth limits much less than glob,
https://stackoverflow.com/questions/55292054/what-is-the-cause-of-the-hard-limit-on-the-directory-nesting-depth-returned-by-g
https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2003/cc781134(v=ws.10)?redirectedfrom=MSDN#maximum-sizes-on-an-ntfs-volume

User will hit these limits before any default we put into glob.  

I'm not worried about false positives _per se_.  Any unexpected results would be files that turned out to match the glob the user specified, even if that's not what the user thought the glob would do.

**Comment by bobhy** [comment]
@jntrnr On the contrary, I can confirm that `{}` are legal in both unix and windows filenames.
But we already had that problem -- glob already supported `[]`, and those are also legal in filenames.

But paths with these special characters to cause a problem:  if you *do* have a path with these glob chars and you specify a similar glob.  Example:
```
> touch a[b]c abc abd
> mkdir foo
> ls
╭───┬───────┬──────┬──────┬──────────╮
│ # │ name  │ type │ size │ modified │
├───┼───────┼──────┼──────┼──────────┤
│ 0 │ a[b]c │ file │  0 B │ now      │
│ 1 │ abc   │ file │  0 B │ now      │
│ 2 │ abd   │ file │  0 B │ now      │
│ 3 │ foo   │ dir  │  2 B │ now      │
╰───┴───────┴──────┴──────┴──────────╯
> ucp a[b]c foo
> ls foo
╭───┬─────────┬──────┬──────┬──────────╮
│ # │  name   │ type │ size │ modified │
├───┼─────────┼──────┼──────┼──────────┤
│ 0 │ foo/abc │ file │  0 B │ now      │
╰───┴─────────┴──────┴──────┴──────────╯
```
User could reasonably complain that the command should have copied file `a[b]c`.

**Comment by bobhy** [comment]
... Looking into this situation a bit more, it's a bigger problem.  Opening #10571 to discuss.

I propose we merge this PR (since it's a step in a direction we want to go and doesn't make the new issue any worse), then focus on resolving the new issue.

**Comment by jntrnr** [comment]
> @jntrnr On the contrary, I can confirm that {} are legal in both unix and windows filenames.

Then let's not go this direction. I'd rather make the built-in glob conservative, and if people want a fancier glob capability, that they would use the `glob` command.

#10498 is not a bug. (I comment on why in the issue)

File: crates/nu-cmd-base/src/arg_glob.rs [hunk 4]
```diff
@@ -9,19 +11,33 @@ const FS_CMD_WALK_BEHAVIOR: WalkBehavior = WalkBehavior {
...
+    pattern: &Spanned<String>, // alleged path or glob
     include_dirs: bool,        // include dirs in results.  Default (f) only include files
     cwd: &PathBuf,             // current working directory
 ) -> Result<Vec<PathBuf>, ShellError> {
-    #[cfg(windows)]
-    let processed_item = &windows_pattern_hack(&pattern.item);
-    #[cfg(not(windows))]
-    let processed_item = &pattern.item;
+    // stat the path first, return path if not not found.
+    match fs::metadata(&pattern.item) {
```
Comment (fdncred) [inline crates/nu-cmd-base/src/arg_glob.rs:24]: oh, nice! this is clever. I like it much better than the windows hack. nice idea.

**Comment by bobhy** [comment]
the last checkin fixes the problem of #10571, but reverts to the problem of absolute glob patterns on Windows: e.g `c:\users\me/*.txt` fails with glob syntax error because both `:` and `\` are glob metacharacters.  But I think I can cover that scenario too.  
Marking as draft while I work on it...

**Comment by bobhy** [comment]
change of plan.  `wax` currently doesn't handle windows paths in several critical ways.  I can't think of a preprocessing scheme that will fix it without a lot of work (and runtime overhead). 

So I'm going to gut this PR and revert to `nu_glob`, then enhance that to support `{}` (and retain the fix for #10571).

Meantime I'll continue discussions with `wax` team to see if we can do something to support windows paths better, but that's a longer term project.

**Comment by bobhy** [comment]
Half way done!

- [x] add  `arg_glob()` wraps `nu_glob::glob_with()` where pattern is file system command argument (and might be a path or a pattern) (fixes #10571)
- [x] convert `ucp` to use `arg_glob()` as proof of concept
- [ ] modify `nu_glob` to support `{}` (match existing files only, not for generating non-existing paths)
- [ ] modify rest of filesystem commands to use `arg_glob()`.

**Comment by bobhy** [comment]
Demo of fix for #10571:

```
> ls
╭────┬────────────┬──────┬──────┬────────────────╮
│  # │    name    │ type │ size │    modified    │
├────┼────────────┼──────┼──────┼────────────────┤
│  0 │ a[b        │ file │  0 B │ a week ago     │
│  1 │ dest       │ dir  │  3 B │ 19 minutes ago │
╰────┴────────────┴──────┴──────┴────────────────╯

> ls dest
╭────────────╮
│ empty list │
╰────────────╯
> cp 'a[b' dest
Error:   × Pattern syntax error near position 30: invalid range pattern
   ╭─[entry #33:1:1]
 1 │ cp 'a[b' dest
   ·    ──┬──
   ·      ╰── invalid pattern
   ╰────

> ucp 'a[b' dest
> ls dest
╭───┬──────────┬──────┬──────┬──────────╮
│ # │   name   │ type │ size │ modified │
├───┼──────────┼──────┼──────┼──────────┤
│ 0 │ dest/a[b │ file │  0 B │ now      │
╰───┴──────────┴──────┴──────┴──────────╯

```

**Comment by bobhy** [comment]
3rd step done!

- [x] add  `arg_glob()` wraps `nu_glob::glob_with()` where pattern is file system command argument (and might be a path or a pattern) (fixes #10571)
- [x] convert `ucp` to use `arg_glob()` as proof of concept
- [ ] modify `nu_glob` to support `{}` (match existing files only, not for generating non-existing paths)
- [x] modify rest of filesystem commands to use `arg_glob()`.

File: crates/nu-cmd-base/Cargo.toml [hunk 5]
```diff
@@ -11,8 +11,15 @@ version = "0.85.1"
...
 [dependencies]
 nu-engine = { path = "../nu-engine", version = "0.85.1" }
+nu-glob = { path = "../nu-glob", version = "0.85.1" }
 nu-parser = { path = "../nu-parser", version = "0.85.1" }
 nu-path = { path = "../nu-path", version = "0.85.1" }
 nu-protocol = { version = "0.85.1", path = "../nu-protocol" }
+nu-utils = { path = "../nu-utils", version = "0.85.1" }
 indexmap = { version = "2.0" }
 miette = { version = "5.10", features = ["fancy-no-backtrace"] }
+wax = { version = "0.5.0", features = ["diagnostics", "miette"] }
```
Comment (fdncred) [inline crates/nu-cmd-base/Cargo.toml:21]:
```suggestion
wax = { version = "0.6.0", features = ["diagnostics", "miette"] }
```
We've recently upgraded the wax crate

**Comment by bobhy** [comment]
Now that I'm down to the bit of the project that brought me here, I'm pretty sure it's a bad idea.
In traditional shells, the `{}` is not really a glob construct, it's what `bash` calls "brace expansion", and it operates entirely at the syntax level, doesn't care whether the resulting strings match files or not.
And that's the feature I want!, so that commands like `mv a.out{,.old}` or `diff {../worktrees/main,.}/crates/nu-cmd-base/src/lib.rs` can be used.   I still think nushell should have this ability, but it's not something that makes sense for a globbing facility to do, and probably should use different metacharacters in deference to nu block `{}`.
`wax` does have a `{}` construct, but it is strictly a pattern matching thing and won't return nonexistant paths (when globbing files), so nu `glob` will support this more limited ability.

So, bottom line, I'm submitting this PR for review as it now stands, updating the title to match what it actually accomplishes and moving on to other work.

**Comment by bobhy** [comment]
Last note: behavior of `ls` command is unchanged, because it uses a different globbing function than `mv`, 'rm`, `cp' and the rest of the filesystem commands.  This might create some user confusion, but I can't wrap my head around it right now.

**Comment by fdncred** [comment]
@bobhy can you update the issue body to explain what exactly this does. there have been quite a few changes and a reverse of direction here. It's difficult to consume this PR as it sits because it's hard to wade through all the subsequent steps you went through to get where you're at here. I'd like to point the core-team here and ask them to make a decision based on what the body/description says and then reviewing the code. Maybe you've already done that. Please let me know.

**Comment by bobhy** [comment]
Squashing the history and reposting as a new clean PR #10694.
